File: internal/cli/verify.go
package cli

import (
	"bufio"
	"database/sql"
	"fmt"
	"io"
	"os"
	"time"
)

// verificationResult holds the counts the verify/doctor commands report. Ports
// the VerificationResult interface (as counts; the TS arrays are only ever used
// by .length).
type verificationResult struct {
	missingArchives         int
	invalidProvenance       int
	missingVectors          int
	orphanVectors           int
	retryableExtractionErrs int
}

// verifyMemoryIndex ports verifyMemoryIndex.
func verifyMemoryIndex(database *sql.DB) (verificationResult, error) {
	var res verificationResult

	rows, err := database.Query(`
		SELECT archive_path, line_start, line_end
		FROM memory_records
		WHERE status = 'active'
		ORDER BY id ASC`)
	if err != nil {
		return res, err
	}
	lineCounts := map[string]int{}
	type rec struct {
		archivePath          string
		lineStart, lineEnd   int
	}
	var records []rec
	for rows.Next() {
		var r rec
		if err := rows.Scan(&r.archivePath, &r.lineStart, &r.lineEnd); err != nil {
			rows.Close()
			return res, err
		}
		records = append(records, r)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return res, err
	}

	for _, r := range records {
		if !pathExists(r.archivePath) {
			res.missingArchives++
			continue
		}
		lineCount, ok := lineCounts[r.archivePath]
		if !ok {
			lineCount = countLines(r.archivePath)
			lineCounts[r.archivePath] = lineCount
		}
		if r.lineStart < 1 || r.lineEnd < r.lineStart || r.lineEnd > lineCount {
			res.invalidProvenance++
		}
	}

	res.missingVectors, err = scalarCount(database, `
		SELECT COUNT(*)
		FROM memory_records m
		LEFT JOIN vec_memory_records v ON v.rowid = m.id
		WHERE m.status = 'active' AND v.rowid IS NULL`)
	if err != nil {
		return res, err
	}

	res.orphanVectors, err = scalarCount(database, `
		SELECT COUNT(*)
		FROM vec_memory_records v
		LEFT JOIN memory_records m ON m.id = v.rowid
		WHERE m.id IS NULL`)
	if err != nil {
		return res, err
	}

	res.retryableExtractionErrs, err = scalarCount(database,
		"SELECT COUNT(*) FROM extraction_state WHERE status = 'errored' AND retry_after <= ?",
		time.Now().UnixMilli())
	if err != nil {
		return res, err
	}

	return res, nil
}

// countLines ports the TS countLines: number of newline-terminated lines, or the
// split count when there is no trailing newline. Returns 0 for an empty file.
func countLines(filePath string) int {
	f, err := os.Open(filePath)
	if err != nil {
		return 0
	}
	defer f.Close()

	r := bufio.NewReader(f)
	var lines, lastByte int = 0, 0
	empty := true
	for {
		b, err := r.ReadByte()
		if err != nil {
			break
		}
		empty = false
		lastByte = int(b)
		if b == '\n' {
			lines++
		}
	}
	if empty {
		return 0
	}
	if lastByte != '\n' {
		lines++ // trailing content without a final newline counts as a line
	}
	return lines
}

// RunVerifyCli runs the `verify` command. Returns a non-nil error (sentinel)
// when issues are found so main can set exit code 1. Ports runVerifyCli.
func RunVerifyCli(out io.Writer) error {
	database, err := openDB()
	if err != nil {
		return err
	}
	defer database.Close()

	v, err := verifyMemoryIndex(database)
	if err != nil {
		return err
	}
	issueCount := v.missingArchives + v.invalidProvenance + v.missingVectors +
		v.orphanVectors + v.retryableExtractionErrs

	if issueCount == 0 {
		fmt.Fprintln(out, "No memory index issues found.")
		return nil
	}

	fmt.Fprintf(out, "Memory index issues: %d\n", issueCount)
	fmt.Fprintf(out, "Missing archives: %d\n", v.missingArchives)
	fmt.Fprintf(out, "Invalid provenance: %d\n", v.invalidProvenance)
	fmt.Fprintf(out, "Missing vectors: %d\n", v.missingVectors)
	fmt.Fprintf(out, "Orphan vectors: %d\n", v.orphanVectors)
	fmt.Fprintf(out, "Retryable extraction errors: %d\n", v.retryableExtractionErrs)
	return ErrVerifyIssues
}

// ErrVerifyIssues is a sentinel indicating verify found issues (exit code 1)
// without an underlying failure.
var ErrVerifyIssues = fmt.Errorf("memory index issues found")

File: plugins/memmem/src/core/verify.ts
import { Database } from 'bun:sqlite';
import { existsSync, readFileSync } from 'fs';

export interface VerificationResult {
  missingArchives: Array<{ id: number; archivePath: string }>;
  invalidProvenance: Array<{ id: number; archivePath: string; lineStart: number; lineEnd: number }>;
  missingVectors: Array<{ id: number; archivePath: string }>;
  orphanVectors: Array<{ rowid: number }>;
  retryableExtractionErrors: Array<{ id: number; archivePath: string; lineStart: number; lineEnd: number }>;
}

interface ActiveMemoryRecord {
  id: number;
  archivePath: string;
  lineStart: number;
  lineEnd: number;
}

function countLines(filePath: string): number {
  const text = readFileSync(filePath, 'utf8');
  if (text.length === 0) return 0;
  return text.endsWith('\n') ? text.split('\n').length - 1 : text.split('\n').length;
}

export function verifyMemoryIndex(db: Database): VerificationResult {
  const missingArchives: VerificationResult['missingArchives'] = [];
  const invalidProvenance: VerificationResult['invalidProvenance'] = [];

  const activeRecords = db.query(`
    SELECT id, archive_path AS archivePath, line_start AS lineStart, line_end AS lineEnd
    FROM memory_records
    WHERE status = 'active'
    ORDER BY id ASC
  `).all() as ActiveMemoryRecord[];

  const lineCounts = new Map<string, number>();
  for (const record of activeRecords) {
    if (!existsSync(record.archivePath)) {
      missingArchives.push({ id: record.id, archivePath: record.archivePath });
      continue;
    }

    let lineCount = lineCounts.get(record.archivePath);
    if (lineCount === undefined) {
      lineCount = countLines(record.archivePath);
      lineCounts.set(record.archivePath, lineCount);
    }

    if (record.lineStart < 1 || record.lineEnd < record.lineStart || record.lineEnd > lineCount) {
      invalidProvenance.push({
        id: record.id,
        archivePath: record.archivePath,
        lineStart: record.lineStart,
        lineEnd: record.lineEnd,
      });
    }
  }

  const missingVectors = db.query(`
    SELECT m.id, m.archive_path AS archivePath
    FROM memory_records m
    LEFT JOIN vec_memory_records v ON v.rowid = m.id
    WHERE m.status = 'active' AND v.rowid IS NULL
    ORDER BY m.id ASC
  `).all() as VerificationResult['missingVectors'];

  const orphanVectors = db.query(`
    SELECT v.rowid
    FROM vec_memory_records v
    LEFT JOIN memory_records m ON m.id = v.rowid
    WHERE m.id IS NULL
    ORDER BY v.rowid ASC
  `).all() as VerificationResult['orphanVectors'];

  const retryableExtractionErrors = db.query(`
    SELECT id, archive_path AS archivePath, line_start AS lineStart, line_end AS lineEnd
    FROM extraction_state
    WHERE status = 'errored' AND retry_after <= ?
    ORDER BY id ASC
  `).all(Date.now()) as VerificationResult['retryableExtractionErrors'];

  return {
    missingArchives,
    invalidProvenance,
    missingVectors,
    orphanVectors,
    retryableExtractionErrors,
  };
}
